src/xpub.cpp

```diff
@@ -177,14 +183,20 @@ void zmq::xpub_t::xread_activated (pipe_t *pipe_)
             //  was removed, or verbose mode is enabled, store it so that
             //  it can be passed to the user on next recv call.
             if (options.type == ZMQ_XPUB && notify) {
-                data = data - 1;
-                size = size + 1;
+                //  Since with inproc we do not store the sub/cancel command
+                //  string anymore in the data buffer, there would be no room
+                //  to hack in the 0/1 byte, and it overwrites the metadata
+                //  pointer in the message struct, so create a new blob_t.
+                //  The pushback makes a copy of the data array anyway, so it
+                //  does not make any difference.
+                blob_t notification (size + 1);
                 if (subscribe)
-                    *data = 1;
+                    *notification.data () = 1;
                 else
-                    *data = 0;
+                    *notification.data () = 0;
+                memcpy (notification.data () + 1, data, size);
 
-                _pending_data.push_back (blob_t (data, size));
+                _pending_data.push_back (ZMQ_MOVE (notification));
```

Comment:
See above. 

Also. this whole part is identical to the part modified above. Can this be extracted into a new member function, so that it can be reused at the two call sites?

Reply:
Given it's an if-else, I'll simply move it out of the block

Reply:
Nice :)